src/librustc_mir/interpret/intern.rs

```diff
@@ -0,0 +1,270 @@
+//! This module specifies the type based interner for constants.
+//!
+//! After a const evaluation has computed a value, before we destroy the const evaluator's session
+//! memory, we need to extract all memory allocations to the global memory pool so they stay around.
+
+use rustc::ty::layout::LayoutOf;
+use rustc::ty;
+use rustc::mir::interpret::{
+    EvalResult, ErrorHandled,
+};
+use rustc::hir::{
+    self,
+    def_id::DefId,
+};
+use super::validity::{RefTracking, PathElem};
+use rustc_data_structures::fx::FxHashSet;
+
+use syntax::ast::Mutability;
+
+use super::{
+    ValueVisitor, MemoryKind, Pointer, AllocKind, AllocId, MPlaceTy, EvalErrorKind,
+};
+use crate::const_eval::{CompileTimeInterpreter, CompileTimeEvalContext};
+
+struct InternVisitor<'rt, 'a: 'rt, 'mir: 'rt, 'tcx: 'a+'rt+'mir> {
+    /// The `path` may be pushed to, but the part that is present when a function
+    /// starts must not be changed!  `visit_fields` and `visit_array` rely on
+    /// this stack discipline.
+    path: Vec<PathElem>,
+    /// previously encountered safe references
+    ref_tracking: &'rt mut RefTracking<MPlaceTy<'tcx>>,
+    ecx: &'rt mut CompileTimeEvalContext<'a, 'mir, 'tcx>,
+    mode: InternMode,
+    /// A list of uninterned but seen relocations to be interned at the end if not interned by
+    /// a safe reference
+    leftover_relocations: &'rt mut FxHashSet<AllocId>,
+}
+
+#[derive(Copy, Clone, Debug)]
+enum InternMode {
+    StaticMut,
+    Static {
+        frozen: bool,
+    },
+    Const {
+        frozen: bool,
+    },
+}
+
+struct IsStaticOrFn;
+
+impl<'rt, 'a, 'mir, 'tcx> InternVisitor<'rt, 'a, 'mir, 'tcx> {
+    fn intern(&mut self, ptr: Pointer) -> EvalResult<'tcx, Option<IsStaticOrFn>> {
+        trace!(
+            "InternVisitor::intern {:?}",
+            ptr,
+        );
+        // remove allocation
+        let tcx = self.ecx.tcx;
+        let memory = self.ecx.memory_mut();
+        let (kind, mut alloc) = match memory.alloc_map.remove(&ptr.alloc_id) {
+            Some(entry) => entry,
+            None => match tcx.alloc_map.lock().get(ptr.alloc_id) {
+                Some(AllocKind::Memory(mem)) => match (self.mode, mem.mutability) {
```

Comment:
These nested matches make my heard hurt.^^

So `self.mode` is for the allocation we are in, `mem.mutability` for the memory it points to?

Reply:
yes